vault/src/vaults.rs

```diff
@@ -35,18 +43,92 @@ impl Vaults {
         let vaults = self.0.read().await;
         vaults.get(&key).cloned()
     }
+
+    /// Returns the position of a given vault in a randomised ordering based on a
+    /// piece of seed data.
+    ///
+    /// # Arguments
+    /// * `data` - the seed used as a basis for the ordering (for example, an issue_id)
+    /// * `account_id` - the AccountId we wish to order relative to the list of vaults
+    pub async fn get_random_position(&self, data: &[u8; 32], account_id: &AccountId) -> usize {
+        fn hash_vault(data: &[u8; 32], account_id: &AccountId) -> sha256::Hash {
+            let account_id: [u8; 32] = account_id.clone().into();
+            let xor = data
+                .zip(account_id) // will need to refactor if we don't want experimental array_zip
+                .map(|(a, b)| a ^ b);
+            sha256::Hash::hash(&xor)
+        }
+
+        let hash = hash_vault(data, account_id);
+        let mut hash_set = HashSet::new(); // for deduping
+        self.0
+            .read()
+            .await
+            .iter()
+            .filter(|&(_, vault_id)| hash_set.insert(hash_vault(data, &vault_id.account_id) < hash))
```

Comment:
Whoops, good spot, yeah this is broken.
If I refactor the trait to hold its own simple list of vaults (as mentioned below) then this will be greatly simplified anyway.